amads/time/meter/__init__.py

```diff
@@ -0,0 +1,619 @@
+    >>> hierarchy.start_hierarchy
+    [[0.0, 4.0], [0.0, 2.0, 4.0]]
+
+    They can be made from a list of `pulse_lengths`
+
+    >>> four_four_from_pulses = MetricalHierarchy(pulse_lengths=[4, 2, 1])
+    >>> four_four_from_pulses.start_hierarchy[0]
+    [0.0, 4.0]
+
+    >>> four_four_from_pulses.start_hierarchy[1]
+    [0.0, 2.0, 4.0]
+
+    >>> four_four_from_pulses.start_hierarchy[2]
+    [0.0, 1.0, 2.0, 3.0, 4.0]
+
+    Those levels of the hierarchy are the same, only the number of levels differs.
+
+    Speaking of levels, they cn be specified by index alongside the time signature string.
+
+    >>> level_test = MetricalHierarchy(time_signature="6/8", levels=[1, 2])
+    >>> level_test.start_hierarchy
+    [[0.0, 3.0], [0.0, 1.5, 3.0], [0.0, 0.5, 1.0, 1.5, 2.0, 2.5, 3.0]]
+
+    Again, for more on these converters, see the static function in this module,
+    e.g., `starts_from_time_signature`, `starts_from_time_signature_and_levels`.
+
+    """
+
+    def __init__(
+        self,
+        time_signature: Optional[str] = None,
+        levels: Optional[list] = None,
+        pulse_lengths: Optional[list] = None,
+        start_hierarchy: Optional[list] = None,
+        names: Optional[dict] = None,
+    ):
+        # Retrieve or create the metrical structure
+        if start_hierarchy:
+            self.start_hierarchy = start_hierarchy
+            return
+        elif pulse_lengths:
+            self.pulse_lengths = pulse_lengths
+            self.start_hierarchy = starts_from_pulse_lengths(
+                pulse_lengths, require_2_or_3_between_levels=False
+            )
+            return
+        elif levels:
+            self.levels = levels
+            if time_signature:  # both time signature and levels
+                self.time_signature = time_signature
+                self.start_hierarchy = starts_from_time_signature_and_levels(
+                    time_signature, levels
+                )
+            else:  # no time signature, yes levels
+                raise ValueError(
+                    "To specify levels, please also enter a valid time signature."
+                )
+        elif time_signature:  # time signature and no levels, assume all
+            self.time_signature = time_signature
+            self.start_hierarchy = starts_from_time_signature(time_signature)
+        else:
+            raise ValueError(
+                "Specify at least one of `start_hierarchy`, `pulse_length` or `time_signature`"
+            )
+
+        if names:
+            for key in names:
+                assert isinstance(key, float)
+                assert isinstance(names[key], str)
+
+
+# ------------------------------------------------------------------------------
+
+
+def starts_from_time_signature(
+    ts_str: str, enforce_2s_3s: bool = True, minimum_pulse: int = 64
+) -> list:
+    """
+    Create a start hierarchy for almost any time signature
```

Comment:
It would be good to explain the underlying heuristics being used here, for example to determine whether 6/4 should be grouped into 3s or 2s. If the heuristics only work for a certain set of time signatures, it would be worth stating that explicitly.

Reply:
Ok. Again, I think the clearest explanation is in the `Examples`. I'll add a comments to this effect and re-work them a little for clarity of this specific point.